cpu-load-mon: Initial commit

nik-sie opened the pull request:
### Maintainer:
- Nikita Kuchinsky <nik.owrt@serenity-island.net>

### Compile tested:
- arch: `MediaTek MT7621 ver:1 eco:3`, model: `MikroTik RouterBOARD 760iGS (hEX S)`, OpenWrt version: `OpenWrt 24.10.0`

### Run tested: 
- arch: `MediaTek MT7621 ver:1 eco:3`, model: `MikroTik RouterBOARD 760iGS (hEX S)`, OpenWrt version: `OpenWrt 24.10.0`
- arch: `Atheros AR7161 rev 2`, model: `Aruba AP-105`, OpenWrt version: `OpenWrt 24.10.0`
- arch: `ARMv8 Processor rev 4`, model: `Routerich AX3000`, OpenWrt version: `OpenWrt 23.05.5`

### Tests done:
- service starts automatically, runs stably
- front-end command executes without any errors

### Description:
Initial commit that adds `cpu-load-mon` architecture-independent utility, which might be useful/convenient tool when aggregating various hardware utilization statistics.

If possible, please grant me commit access to let me maintain this utility.

GeorgeSapkin commented:
You should squash your commits, not to pollute the history.

It's a good idea to add a short description of the new package to the initial commit.

Commits must be signed with a real name and match the commit author as per [contributing guidelines](https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md#commits-in-your-pull-requests-should).

GeorgeSapkin commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,44 @@
+endef
+
+define Build/Prepare
+	sed -f ./files/strip-comments.sed $(BUILD_DIR)/cpu-load-avg.awk > $(PKG_BUILD_DIR)/cpu-load-avg.awk
```
Where is `strip-comments.sed` coming from? I don't see it in files.

nik-sie replied:
Omg, I forgot to include it in the commit, my apologies. I'm actually surprised it got build without any errors, when I tested it.

GeorgeSapkin commented on utils/cpu-load-mon/files/cpu-load-mon.init:
```diff
@@ -0,0 +1,15 @@
+
+start_service() {
+	procd_open_instance
+	# procd_set_param env CPU_LOAD_STATS_PATH=/custom/path/to/stats.log
+	procd_set_param command /bin/sh -ec "while true; do awk -f /usr/libexec/cpu-load-stats-log.awk; sleep 60; done"
+	procd_set_param term_timeout 1 # Specifies the amount of seconds to wait for a clean process exit after delivering the TERM signal. Default: 5
+	procd_set_param respawn 180 10 10 # threshold of 180 with a retry value of 10 will cause procd to consider the service to be crashed if the associated UNIX process terminated more than 10 times within a time frame of 3 minutes.
```
Any reason why these are not configurable through a config file?

nik-sie replied:
Regarding the `CPU_LOAD_STATS_PATH`, yes I agree, this one might be specified via config, I just have not yet fully implemented and tested it.
regarding the rest, did not see it necessary, since I do not expect those to change and do not want to bother users configuring them.
I just found that default values of `term_timeout = 5` and `respawn = 3600 5 5` are too high for the given functionality. Also, examples I studied never specified those via config.

BKPepe commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,44 @@
+PKG_MAINTAINER:=Nikita Kuchinsky <nik.owrt@serenity-island.net>
+
+PKG_SOURCE=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_VERSION:=$(PKG_VERSION)
```
Is this used somewhere?

nik-sie replied:
I use it in all of my 40 routers and couple Debian servers to track hardware utilization. This is my first attempt to contribute to OpenWRT, so I wanted to start with something simple.

GeorgeSapkin replied:
I think the question might have been about the `PKG_SOURCE_VERSION` variable.

nik-sie replied:
Ah, thank you for clarifying, yes, it is used by both, the service and the front-end utility. I am still thinking about the way to customize path for the stats file conveniently and efficiently; I would like to avoid reading and parsing some kind of config file on every single call to this simple script.

GeorgeSapkin replied:
I think `PKG_SOURCE_VERSION` is used to define `PKG_VERSION` and `PKG_SOURCE` as per: https://github.com/openwrt/openwrt/blob/f44984f19c447db30b76ca7f49572aac97270145/include/download.mk#L11-L17

```Makefile
ifdef PKG_SOURCE_VERSION
  ifndef PKG_VERSION
    PKG_VERSION := $(if $(PKG_SOURCE_DATE),$(subst -,.,$(PKG_SOURCE_DATE)),0)~$(call version_abbrev,$(PKG_SOURCE_VERSION))
  endif
  PKG_SOURCE_SUBDIR ?= $(PKG_NAME)-$(PKG_VERSION)
  PKG_SOURCE ?= $(PKG_SOURCE_SUBDIR).tar.zst
endif
```

Both of which are already defined. So it seems redundant.

nik-sie replied:
Agree, thank you for pointing that. Fixed.

GeorgeSapkin commented:
You should rebase/fast-forward on the target branch, instead of creating a merge commit. This just creates noise in the history and will prevent this PR from being merged.

nik-sie commented:
My apologies, still figuring out the workflow. Will do ASAP

GeorgeSapkin commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,53 @@
+SRC_DIR:=$(PKG_BUILD_DIR)/$(PKG_NAME)
+define Build/Prepare
+	$(call Build/Prepare/Default)
+	sed -f ./files/strip-comments.sed $(SRC_DIR)/cpu-load-avg.awk       > $(PKG_BUILD_DIR)/cpu-load-avg.awk
```
How big are those comments if they need to be stripped?

nik-sie replied:
There isn't too much, 4.6KB with VS 0.9KB. But would still like to save as much space as possible on OpenWRT while keeping comments for other operating systems, e.g. Debian

GeorgeSapkin commented on utils/cpu-load-mon/files/cpu-load-mon.init:
```diff
@@ -0,0 +1,20 @@
+	if [ -n "$val" ]; then procd_set_param env CPU_LOAD_STATS_PATH=$val; fi
+	if [ ! -d /var/run ]; then mkdir -p /var/run; fi
+	if [ ! -x /run ]; then ln -s /var/run /run; fi
+	procd_set_param command /bin/sh -ec "while true; do awk -f /usr/libexec/cpu-load-stats-log.awk; sleep 60; done"
```
Could this be a cron job?

nik-sie replied:
Oh, certainly, that's exactly how I run it on Debian, just not sure how to implement safe adding/removing cron jobs on OpenWRT. Editing from script `/etc/crontabs/root` sounds like a bit dangerous to me and I haven't found a better way yet.

GeorgeSapkin replied:
Not sure if this is the best way, but it's there: https://github.com/openwrt/packages/blob/master/net/acme-common/files/acme.init#L144-L146

```bash
start_service() {
	mkdir -p $run_dir
	mkdir -p "$CHALLENGE_DIR"

	grep -q '/etc/init.d/acme' /etc/crontabs/root 2>/dev/null || {
		echo "0 0 * * * /etc/init.d/acme renew" >>/etc/crontabs/root
	}
}
```

nik-sie replied:
Yes, adding is quite straightforward, but removing on uninstall is a bit trickier...
There is also an option to add dependency on something like `micrond`: _Small Cron daemon providing a cron.d directory other packages can install their crontabs into_. But I would like to avoid dependencies for such a basic set of scripts.

GeorgeSapkin commented:
You should name your commit something like: `cpu-load-mon: add new project`.

And it's customary for commit messages to follow the 50/72 line length, unless unavoidable.

https://github.com/openwrt/packages/blob/master/CONTRIBUTING.md#commits-in-your-pull-requests-should

GeorgeSapkin commented on utils/cpu-load-mon/files/cpu-load-mon.init:
```diff
@@ -0,0 +1,20 @@
+
+start_service() {
+	procd_open_instance
+	config_load "cpu-load-mon"
```
If you're using a config, it's a good idea to have a default one with all the potential options described.

nik-sie replied:
Agree, fixed.

GeorgeSapkin commented on utils/cpu-load-mon/files/cpu-load-mon.init:
```diff
@@ -0,0 +1,20 @@
+	local val
+	config_get val "stats_path" ""
+	if [ -n "$val" ]; then procd_set_param env CPU_LOAD_STATS_PATH=$val; fi
+	if [ ! -d /var/run ]; then mkdir -p /var/run; fi
+	if [ ! -x /run ]; then ln -s /var/run /run; fi
```
Is it ever the case that these would not exist?

nik-sie replied:
Yes, on OpenWRT 23.05 there is no `/run` link. Some other packages also unconditionally call `mkdir -p /var/run`.

BKPepe replied:
Why we should care about OpenWrt 23.05 if it is outdated and not supported?

GeorgeSapkin commented on utils/cpu-load-mon/files/cpu-load-mon.init:
```diff
@@ -0,0 +1,20 @@
+start_service() {
+	procd_open_instance
+	config_load "cpu-load-mon"
+	local val
```
I suppose this could be something else than `val`, like `stats_path`.

nik-sie replied:
Well, yes, but it's completely irrelevant here, since this is just a temporary variable that holds result of `config_get` command/macro. I might later re-use this variable to acquire value for customizing frequency of calling `cpu-load-stats-log.awk` in the cycle if it ever becomes necessary.

GeorgeSapkin commented on utils/cpu-load-mon/files/cpu-load-mon.init:
```diff
@@ -0,0 +1,20 @@
+	config_load "cpu-load-mon"
+	local val
+	config_get val "stats_path" ""
+	if [ -n "$val" ]; then procd_set_param env CPU_LOAD_STATS_PATH=$val; fi
```
You should double-quote the `$val` when assigning it: https://www.shellcheck.net/wiki/SC2086

nik-sie replied:
Uh... my bad, thank you for point that out.

Commit:
cpu-load-mon: add new package
Description: CPU Load Monitor service and utility written in awk (mawk)
Gathers CPU utilization data over time in the background and displays
averages over the last 1 min, 5 min, 15 min and overall (since boot).

Signed-off-by: Nikita Kuchinsky <nik.owrt@serenity-island.net>

BKPepe commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,53 @@
+	TITLE:=CPU Load Monitor
+	URL:=https://codeberg.org/sie-foss/cpu-load-mon
+	PKGARCH:=all
+	PROVIDES:=cpu-load-avg
```
I am not sure if the PROVIDES here is right. Are you sure about it?

BKPepe commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,53 @@
+PKG_NAME:=cpu-load-mon
+PKG_VERSION=1.0
+PKG_RELEASE:=1
+MAINTAINER:=Nikita Kuchinsky <nik.owrt@serenity-island.net>
```
```suggestion
PKG_MAINTAINER:=Nikita Kuchinsky <nik.owrt@serenity-island.net>
```

BKPepe replied:
We also missing here ``PKG_LICENSE``, ``PKG_LICENSE_FILES``

BKPepe commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,53 @@
+
+include $(INCLUDE_DIR)/package.mk
+
+define Package/cpu-load-mon
```
No depends?

BKPepe commented on utils/cpu-load-mon/files/strip-comments.sed:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sed -f
+
+1{/^#!/ {p}}; /^[\t\ ]*#/d;/\.*#.*/ {/[\x22\x27].*#.*[\x22\x27]/ !{:regular_loop s/[\t ]*[^\]#.*//;t regular_loop}; /[\x22\x27].*#.*[\x22\x27]/ {:special_loop s/\([\x22\x27].*#.*[^\x22\x27]\)#.*/\1/;t special_loop}; /\\#/ {:second_special_loop s/\(.*\\#.*[^\]\)#.*/\1/;t second_special_loop}}
```
Hm... This requires a lot of time to read it. 🤷

BKPepe commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,53 @@
+averages over the last 1 min, 5 min, 15 min and overall (since boot).
+endef
+
+UNPACK_CMD=tar -xaf "$(DL_DIR)/$(PKG_SOURCE)" --directory $(PKG_BUILD_DIR)
+SRC_DIR:=$(PKG_BUILD_DIR)/$(PKG_NAME)
+define Build/Prepare
+	$(call Build/Prepare/Default)
+	sed -f ./files/strip-comments.sed $(SRC_DIR)/cpu-load-avg.awk       > $(PKG_BUILD_DIR)/cpu-load-avg.awk
+	sed -f ./files/strip-comments.sed $(SRC_DIR)/cpu-load-stats-log.awk > $(PKG_BUILD_DIR)/cpu-load-stats-log.awk
+	sed -f ./files/strip-comments.sed ./files/cpu-load-mon.init         > $(PKG_BUILD_DIR)/cpu-load-mon.init
+	$(Build/Patch)
+endef
```
O_o.  But you can do it differently, right? This is not how it should work, though.

BKPepe commented on utils/cpu-load-mon/Makefile:
```diff
@@ -0,0 +1,53 @@
+
+define Package/cpu-load-mon/install
+	$(INSTALL_DIR) $(1)/usr/bin $(1)/usr/libexec $(1)/etc/init.d $(1)/etc/config
+	$(INSTALL_DATA) ./files/cpu-load-mon.conf $(1)/etc/config/$(PKG_NAME)
```
```suggestion
	$(INSTALL_CONF) ./files/cpu-load-mon.conf $(1)/etc/config/cpu-load-mon
```

Why this file is not marked as configuration file?

BKPepe reviewed (changes requested):
Looking at the repository on codeberg, the project basically started and died on the exact same day – 6 commits, all labeled as 'Initial commit', and not a peep ever since. I might be a bit too harsh, but I'm closing this pull request. :)

We have other, maintained alternatives for CPU measurement here.
